x/concentrated-liquidity/swaps.go

```diff
@@ -660,39 +673,52 @@ func (k Keeper) updatePoolForSwap(
 	// Fixed gas consumption per swap to prevent spam
 	poolId := pool.GetId()
 	ctx.GasMeter().ConsumeGas(gammtypes.BalancerGasFeeForSwap, "cl pool swap computation")
-	pool, err := k.getPoolById(ctx, poolId)
-	if err != nil {
-		return err
-	}
 
 	// Spread factors should already be rounded up to a whole number dec, but we do this as a precaution
 	spreadFactorsRoundedUp := sdk.NewCoin(swapDetails.TokenIn.Denom, totalSpreadFactors.Ceil().TruncateInt())
 
 	// Remove the spread factors from the input token
 	swapDetails.TokenIn.Amount = swapDetails.TokenIn.Amount.Sub(spreadFactorsRoundedUp.Amount)
 
-	// Send the input token from the user to the pool's primary address
-	err = k.bankKeeper.SendCoins(ctx, swapDetails.Sender, pool.GetAddress(), sdk.Coins{
-		swapDetails.TokenIn,
-	})
+	// InputOutputCoins performs multi-send functionality. It accepts a series of
+	// inputs that correspond to a series of outputs. It returns an error if the
+	// inputs and outputs don't lineup or if any single transfer of tokens fails.
+	err := k.bankKeeper.InputOutputCoins(ctx, []banktypes.Input{{
+		Address: swapDetails.Sender.String(),
+		Coins:   sdk.NewCoins(swapDetails.TokenIn),
+	}}, []banktypes.Output{
+		{
+			Address: pool.GetAddress().String(),
+			Coins:   sdk.NewCoins(swapDetails.TokenIn),
+		}})
 	if err != nil {
 		return types.InsufficientUserBalanceError{Err: err}
 	}
 
 	// Send the spread factors taken from the input token from the user to the pool's spread factor account
 	if !spreadFactorsRoundedUp.IsZero() {
-		err = k.bankKeeper.SendCoins(ctx, swapDetails.Sender, pool.GetSpreadRewardsAddress(), sdk.Coins{
-			spreadFactorsRoundedUp,
-		})
+		err := k.bankKeeper.InputOutputCoins(ctx, []banktypes.Input{{
+			Address: swapDetails.Sender.String(),
+			Coins:   sdk.NewCoins(spreadFactorsRoundedUp),
+		}}, []banktypes.Output{
+			{
+				Address: pool.GetSpreadRewardsAddress().String(),
+				Coins:   sdk.NewCoins(spreadFactorsRoundedUp),
+			}})
 		if err != nil {
 			return types.InsufficientUserBalanceError{Err: err}
 		}
 	}
 
-	// Send the output token to the sender from the pool
-	err = k.bankKeeper.SendCoins(ctx, pool.GetAddress(), swapDetails.Sender, sdk.Coins{
-		swapDetails.TokenOut,
-	})
+	err = k.bankKeeper.InputOutputCoins(ctx, []banktypes.Input{{
+		Address: pool.GetAddress().String(),
+		Coins:   sdk.NewCoins(swapDetails.TokenOut),
+	}}, []banktypes.Output{
+		{
+			Address: swapDetails.Sender.String(),
+			Coins:   sdk.NewCoins(swapDetails.TokenOut),
+		}})
```

Comment:
What is the benefit of using multi-send API when we replace each regular API with a multi-send one?

I would assume that we would want to batch all inputs and all outputs and run a single multi-send API at the end.

Reply:
yea i think so, using InputOutputCoins made gas slightly lower. ref: https://github.com/osmosis-labs/osmosis/pull/5850/files#r1266377932

Reply:
This change here isn't quite what I had in mind. I was expecting to be doing a MultiSend (not an InputOutputCoins!) and combining the two sends from the sender! My bad, I forgot that MultiSend is the name of the SDK message, which doesn't actually use the sendManyCoins API.

https://github.com/osmosis-labs/cosmos-sdk/blob/osmosis-main/x/bank/keeper/send.go#L21C7-L21C7

Reply:
Though this suggests an interesting idea, thats actually better than mine.

Could we do all the sends here as a single InputOutputCoins call?

Reply:
This still does not make sense to me, this still looks like three separate bank sends to me.